### DISCUSSION TITLE
Dynamic svg icons

### DISCUSSION THREAD

**Discussion by gitFoxCode** [nuxt/nuxt#17606] [Questions]
Hello, I want to add icons to my project. I would like all .svg files from  'assets/icons' folder to be automatically usable in components using `<icon name="svg-file-name-from-icons-folder" />`. I would like these icons to be "font-like", meaning that you can manipulate their size using `font-size` and that you can change their color using `color`.
 Additionally, in case the svg icon has a gradient I would like to be able to keep the original color from the file. I would like to get an effect similar to this: [https://css-tricks.com/a-font-like-svg-icon-system-for-vue/](https://css-tricks.com/a-font-like-svg-icon-system-for-vue/)
I tried to create something like this using `const icons = import.meta.glob('../assets/icons/**.svg')` but there were problems in production and errors during nuxt's hot-reload (I have to refresh the page twice to load all icons properly). 
How do you use icons in your projects and what is the best practice in this case?

**Comment by sewalsh** [comment]
If the icons aren't bespoke designed ones by you, it sounds like you should check out [unplugin-icons](https://github.com/antfu/unplugin-icons)

**Reply by gitFoxCode** [reply]
I wouldn't want to be limited to pre-made icons

**Comment by yunyuyuan** [comment]
My way to dynamic load sprite-svg-icon on-demand,you can set `font-size` and `fill` to it.

### In `nuxt.config.ts`
```ts
import fs from "fs";
import { defineNuxtConfig } from "nuxt";
import { Plugin } from "vite";
import { dataToEsm } from "rollup-pluginutils";

const rawLoaderPlugin: Plugin = {
  name: "raw-file-loader",
  transform (_: string, filepath: string) {
    if (filepath.includes("node_modules")) {
      return null;
    }
    if (filepath.endsWith(".svg")) {
      return {
        code: dataToEsm(fs.readFileSync(filepath).toString())
      };
    }
  }
};

export default defineNuxtConfig({
  vite: {
    plugins: [rawLoaderPlugin]
  }
})
```
### In `plugins/svg.client.ts`
```ts
export default defineNuxtPlugin(() => {
  const svgContainer = document.createElement("div");
  svgContainer.id = "_SVG_SPRITE_CONTAINER_";
  svgContainer.setAttribute("xmlns", "http://www.w3.org/2000/svg");
  svgContainer.setAttribute(
    "style",
    "position: absolute; width: 0; height: 0;overflow: hidden"
  );
  document.body.appendChild(svgContainer);
});
```
### In `components/svg-icon.vue`
```vue
<script setup lang="ts">
const props = defineProps({
  name: { type: String, required: true }
});

const icon = ref<string>("");

if (process.client) {
  watch(
    () => props.name,
    (name) => {
      import(`../assets/svg/${name}.svg`).then((res) => {
        const id = `icon-${name}`;
        icon.value = "#" + id;
        const container = document.getElementById("_SVG_SPRITE_CONTAINER_");
        if (!container.querySelector(icon.value)) {
          const svgElement = new DOMParser()
            .parseFromString(res.default, "image/svg+xml")
            .querySelector("svg");
          if (svgElement) {
            for (const key of ["width", "height", "x", "y"]) {
              svgElement.removeAttribute(key);
            }
            svgElement.id = id;
            container.appendChild(svgElement);
          }
        }
      });
    },
    { immediate: true }
  );
}
</script>

<template>
  <svg class="--icon-svg common-svg" aria-hidden="true">
    <use :href="icon" />
  </svg>
</template>
```

Then put your svg files in `~/assets/svg/my-svg-icon.svg`,use it just like `<svg-icon name="my-svg-icon"/>`,and `my-svg-icon.svg` will be loaded dynamiclly.

**Reply by gitFoxCode** [reply]
Since creating this thread, I have also started working on a separate icon module myself to make working with own icons even easier 
https://github.com/gitFoxCode/nuxt-icons
